### ISSUE TITLE
[LightGBM] [Fatal] Model file doesn't contain feature names

### ISSUE THREAD

**Issue by heroxrq** [Microsoft/LightGBM#393]
## Environment info
Operating System: CentOS Linux release 7.2.1511
CPU: Intel(R) Xeon(R) CPU E5-2650 v3 @ 2.30GHz
C++/Python/R version: 4.8.5

## Error Message:
[LightGBM] [Fatal] Model file doesn't contain feature names

## Reproducible examples
the data is libsvm format
max_idx 9250421
distinct_idx 3492428

when I reduce the number of features to 500000, it is ok, but if increase the feature to 5 millions or 10 millions, it report the Fatal log.

## Steps to reproduce
I added some log in the c++ src files, and here are some shotcuts:

  // get feature names
  line = Common::FindFromLines(lines, "feature_names=");
  std::cout << "feature name line: " << line << std::endl;
  std::cout << "line size: " << line.size() << std::endl;
  if (line.size() > 0) {
    feature_names_ = Common::Split(line.substr(std::strlen("feature_names=")).c_str(), " ");
    if (feature_names_.size() != static_cast<size_t>(max_feature_idx_ + 1)) {
      Log::Fatal("Wrong size of feature_names");
      return false;
    }
  } else {
    Log::Fatal("Model file doesn't contain feature names");
    return false;
  }

Load dataset...
Start training...
num_total_features: 9249959
used_features: 332536
num_features: 332536
num_groups: 332536
[LightGBM] [Info] Number of positive: 2286045, number of negative: 1644747
[LightGBM] [Info] Total Bins 665072
[LightGBM] [Info] Number of data: 3930792, number of used features: 332536

lightgbm config=predict.conf          
[LightGBM] [Info] Finished loading parameters
[LightGBM] [Info] Warning: last line of model.txt has no end of line, still using this line
feature name line: 
line size: 0
[LightGBM] [Fatal] Model file doesn't contain feature names
Met Exceptions:
Model file doesn't contain feature names

**Comment by guolinke** [comment]
Did you use cpp version? and perform continue training?
BTW,  what is the content in the "model.txt" ? Did you use "model.txt" as both input_model and output_model ?

**Comment by heroxrq** [comment]
I have tried using both cli or python to train, and use the cli to load the model.txt file to predict.

I didnot perform continue training.
The model.txt is the output of training, I didnot use it as input_model.

I seems that when the number of features is big, the cli predict does not work.
----------------------------------------------------------------------------------------------
feature_num     train         predict_exp1    is_ok   predict_exp2    is_ok
500000             cli             cli                     yes
500000             python     python             yes      cli                     yes
5000000           cli             cli                     no
5000000           python     cli                     no       python             yes
----------------------------------------------------------------------------------------------
here is python code, BTW, I didnot reload the model.txt in python code to see if it can also predicts, this need to be tried later :

print('Load dataset...')
lgb_train = lgb.Dataset('binary.train')
lgb_valid = lgb.Dataset('binary.valid')
lgb_test = lgb.Dataset('binary.test')

# specify your configurations as a dict
params = {
    'task': 'train',
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'learning_rate': 0.1,
    'num_leaves': 63,
    'feature_fraction': 0.9,
    'bagging_fraction': 0.8,
    'bagging_freq': 5,
    'max_bin': 255,
    'metric': {'binary_logloss', 'binary_error', 'auc'},
    'metric_freq': 1,
    'is_training_metric': True
}

print('Start training...')
# train
gbm = lgb.train(params,
                lgb_train,
                num_boost_round=2,
                valid_sets=[lgb_valid, lgb_test],
                early_stopping_rounds=5)

print('Save model...')
# save model to file
gbm.save_model('model.txt')

print('Start predicting...')
# predict
y_pred = gbm.predict('binary.test', num_iteration=gbm.best_iteration)
print y_pred[: 1000]

----------------------------------------------------------------------------------------------
here is the head lines of model.txt, the features are too many, I just list some.

tree
num_class=1
num_tree_per_iteration=1
label_index=0
max_feature_idx=9249958
objective=binary sigmoid:1
feature_names=Column_0 Column_1 Column_2 Column_3 Column_4 Column_5 Column_6 Column_7 Column_8 Column_9 Column_10 Column_11 Column_12 Column_13 Column_14 Column_15 Column_16 Column_17 Column_18 Column_19 Column_20 Column_21 Column_22 Column_23 Column_24 Column_25 Column_26 Column_27 Column_28 Column_29 Column_30 Column_31 Column_32 Column_33 Column_34 Column_35 Column_36 Column_37 Column_38 Column_39 Column_40 Column_41............

**Comment by heroxrq** [comment]
$ less model.txt
tree
num_class=1
num_tree_per_iteration=1
label_index=0
max_feature_idx=9249958
objective=binary sigmoid:1
feature_names=Column_0 Column_1 Column_2 Column_3 Column_4 Column_5 Column_6 Column_7 Column_8 Column_9 Column_10 Column_11 Column_12 Column_13 Column_14 Column_15 Column_16 Column_17 Column_18 Column_19 Column_20 Column_21 Column_22 Column_23 Column_24 Column_25 Column_26 Column_27 Column_28 Column_29 Column_30 Column_31 Column_32 Column_33 Column_34 Column_35 Column_36 Column_37 Column_38 Column_39 Column_40 Column_41............
------------------------------------------------------------------------------------------
I output the model_str in GBDT::LoadModelFromString in the following code:
bool GBDT::LoadModelFromString(const std::string& model_str) {
  std::ofstream out_file("model_str.txt", std::ofstream::out);
  out_file << model_str << std::endl;
  out_file.close();
  // use serialized string to restore this object
  models_.clear();
  std::vector<std::string> lines = Common::Split(model_str.c_str(), '\n');

However, "model_str.txt" doesnot contain "feature_names="
------------------------------------------------------------------------------------------
$less model_str.txt
num_class=1
num_tree_per_iteration=1
label_index=0
max_feature_idx=9249958
objective=binary sigmoid:1
433 Column_7903434 Column_7903435 Column_7903436 Column_7903437 Column_7903438 Column_7903439 Column_7903440 Column_7903441 Column_7903442 Column_7903443 Column_7903444 Column_7903445 Column_7903446 Column_7903447.........

**Comment by guolinke** [comment]
@heroxrq Thanks,
I think there may be some problem for large line in the reader .

**Comment by heroxrq** [comment]
I think the loading of model exists some bugs when the "feature_names" line is long.

**Comment by guolinke** [comment]
@heroxrq can you give me this model file that I can debug on it ?

**Comment by heroxrq** [comment]
ok, great! The model.txt is 176M, which way does I git it to you? email?

**Comment by guolinke** [comment]
@heroxrq OK

**Comment by heroxrq** [comment]
@guolinke 
I sent the model.txt and 10 lines of testset file to the mail address "i@yumumu.me", please check it.

**Comment by guolinke** [comment]
fixed it in the latest code.
